core/include/keyman/keyman_core_api.h

```diff
@@ -526,6 +558,152 @@ enum km_core_action_type {
   KM_CORE_IT_MAX_TYPE_ID
 };
 
+/*
+```
+### Actions
+This structure provides the results of processing a key event to the Platform layer and
+should be processed by the Platform layer to issue commands to the os text
+services framework to transform the text store in the Client Application, among
+other actions.
+
+This API replaces the Action items APIs, which is now deprecated and will be
+removed in the future.
+```c
+*/
+
+typedef enum { KM_CORE_FALSE = 0, KM_CORE_TRUE = 1 } km_core_bool;
+typedef enum { KM_CORE_CAPS_UNCHANGED = -1, KM_CORE_CAPS_OFF = 0, KM_CORE_CAPS_ON = 1 } km_core_caps_state;
+
+typedef struct {
+  // number of codepoints (not codeunits!) to delete from app context.
+  unsigned int code_points_to_delete;
+
+  // null-term string of characters to insert into document
+  km_core_usv* output;
+
+  // list of options to persist, terminated with KM_CORE_OPTIONS_END
```

Comment:
```suggestion
  // list of options to persist, terminated with KM_CORE_OPTIONS_END, or null
```

Reply:
Hmm, I want them to use KM_CORE_OPTIONS_END so won't encourage use of null :grin:

Reply:
The way it's documented I'd expect to have an array with just one element, KM_CORE_OPTIONS_END, if there are no options, but that doesn't match the implementation. So I think we should document that this can be null if there are no options to persist. Especially since `output` above will be an empty string (instead of null) if there is no output.

Reply:
Gotcha. I think I will change `options` to always point to a valid array ending in `KM_CORE_OPTIONS_END`. That's one less test for the consumer.